src/rules/consentric-integration.js

```diff
@@ -0,0 +1,184 @@
+    };
+
+    // Creates Citizen Record in Consentric with Auth0 Id
+    const createCitizen = ({ userRef, apiAccessToken }) => {
+        console.log(`Upserting Consentric Citizen record for ${userRef}`);
+        const data = {
+            applicationId: CONSENTRIC_APPLICATION_ID,
+            externalRef: userRef,
+        };
+
+        return consentricApi
+            .post('/v1/citizens', data, {
+                headers: {
+                    Authorization: 'Bearer ' + apiAccessToken
+                },
+            })
+            .catch(err => {
+                if (err.response.status !== 409) { // 409 indicates Citizen with given reference already exists in Consentric
+                    console.error(err);
+                    throw err;
+                }
+            });
+    };
+
+    // Function to retrieve Consentric User Token from User Metadata
+    const getConsentricUserTokenFromMetadata = user => user.app_metadata && user.app_metadata.consentric;
+
+    // Generates On Demand Consentric User Token for the given User using the API Access Token
+    const generateConsentricUserAccessToken = async ({ userRef, apiAccessToken }) => {
+        try {
+            console.log(`Attempting to generate access token API for ${userRef}`);
+
+            const { data: { token, expiryDate: exp } } = await consentricApi
+                .post('/v1/access-tokens/tokens',
+                    {
+                        applicationId: CONSENTRIC_APPLICATION_ID,
+                        externalRef: userRef,
+                        expiryDate: moment().add(3, 'months').toISOString()
+                    },
+                    {
+                        headers: {
+                            Authorization: 'Bearer ' + apiAccessToken
+                        }
+                    }
+                );
+
+            return {
+                token,
+                exp
+            };
+
+        } catch (err) {
+            console.error(err);
+            throw err;
+        }
+    };
+
+    const loadConsentricUserAccessToken = async ({ user }) => {
+        try {
+            const metadataUserToken = getConsentricUserTokenFromMetadata(user);            
+            if ((metadataUserToken) && moment(metadataUserToken.exp).subtract(1, "days").isAfter(moment())) return metadataUserToken;
+        
+            const { jwt: apiAccessToken } = await getConsentricApiAccessToken();            
+            const apiCredentials = {
+                userRef: user.user_id,
+                apiAccessToken,
+            };
+
+            // Create Citizen with Auth0 UserId
+            await createCitizen(apiCredentials);
+
+            // Generate an On Demand Access Token for the created citizen
+            const generatedToken = await generateConsentricUserAccessToken(apiCredentials);
+
+            // Persist the app_metadata update            
+            await auth0.users.updateAppMetadata(user.user_id, { consentric: generatedToken });
```

Comment:
Something to keep in mind here ... this call is constrained by the tenant's rate limits so if it's running on every login, there is effectively a 10 login/second limit for free accounts and 50 login/second for paid accounts:

https://auth0.com/docs/policies/rate-limit-policy/management-api-endpoint-rate-limits

Since we're using a stored token if it's there then this is less of an issue down the road but it would be worth mentioning this limit in your integration guide as a sudden spike of new users could fail pretty quickly.

Reply:
I've asked Dom to update the guide with an appropriate warning :)